migrations/2018-08-28-010336_init/up.sql

```diff
@@ -0,0 +1,70 @@
+CREATE DATABASE IF NOT EXISTS `syncstorage` /*!40100 DEFAULT CHARACTER SET latin1 */;
+
+USE `syncstorage`;
+
+-- DROP TABLE IF EXISTS `bso`;
+CREATE TABLE `bso` (
+    `user_id` INT                           NOT NULL,
+    `collection_id` INT                     NOT NULL,
+    `id` VARCHAR(64)                        NOT NULL,
+
+    `sortindex` INT DEFAULT NULL,
+
+    `payload` MEDIUMTEXT                    NOT NULL,
+    `payload_size` INT DEFAULT '0'          NOT NULL,
+
+    -- last modified time in milliseconds since epoch
+    `modified` BIGINT                       NOT NULL,
+    -- expiration in seconds milliseconds since epoch
+    `expiry` BIGINT DEFAULT '3153600000000' NOT NULL,
+
+    PRIMARY KEY (`user_id`, `collection_id`, `id`),
+    KEY `bso_expiry_idx` (`expiry`),
+    KEY `bso_usr_col_mod_idx` (`user_id`, `collection_id`, `modified`)
+) ENGINE=InnoDB DEFAULT CHARSET=latin1;
+
+
+-- DROP TABLE IF EXISTS `collections`;
+CREATE TABLE `collections` (
+    `id` INT PRIMARY KEY          NOT NULL AUTO_INCREMENT,
+    `name` VARCHAR(32) UNIQUE     NOT NULL,
+    `modified` BIGINT DEFAULT '0' NOT NULL
+) ENGINE=InnoDB AUTO_INCREMENT=100 DEFAULT CHARSET=latin1;
+INSERT INTO collections (id, name) VALUES
+    ( 1, "clients"),
+    ( 2, "crypto"),
+    ( 3, "forms"),
+    ( 4, "history"),
+    ( 5, "keys"),
+    ( 6, "meta"),
+    ( 7, "bookmarks"),
+    ( 8, "prefs"),
+    ( 9, "tabs"),
+    (10, "passwords"),
+    (11, "addons"),
+    (12, "addresses"),
+    (13, "creditcards");
+
+
+-- DROP TABLE IF EXISTS `user_collections`;
+CREATE TABLE `user_collections` (
+    `user_id` INT       NOT NULL,
+    `collection_id` INT NOT NULL,
+    `modified` BIGINT   NOT NULL,
```

Comment:
This should be the last-modified timestamp of the collection as a whole, and should change whenever an item is added, modified or deleted in the collection. It will be greater-than-or-equal-to the MAX() of the modified column for that collection in the bso table, with the greater-than case being because an item was deleted.

FWIW, the python server has a bug here in its handling of deleted collections:

  https://github.com/mozilla-services/server-syncstorage/pull/62

Basically, it tries to calculate the last-modified time of the storage as a whole by doing `SELECT MAX(modified) FROM user_collections WHERE uid = X`.  That's incorrect in the case of a deleted collection, which should cause the last-modified time of the storage as a whole to increase, but won't affect the last-modified time of any remaining collections.

It's clearly an edge-case, because we haven't bothered to actually fix it in the python version.  But for greenfields code it's probably worth doing it right the first time.  I suggest adding a `deleted_at` timestamp field to the `user_collections` table, so that we can explicit track the deletion of collections.  But that can be a follow-up issue if necessary.

Reply:
https://github.com/mozilla-services/syncstorage-rs/issues/23